File: backend/app/services/numbering.py
"""Human-friendly complaint numbers, e.g. ``RW-2026-001024``.

The UUID primary key is what the system uses internally; this is what a citizen
reads over the phone. Numbers restart each calendar year.

The sequence is derived from the highest number already issued this year and
written under the table's unique constraint, so a collision from two concurrent
submissions surfaces as an ``IntegrityError`` and is retried rather than
silently issuing a duplicate.
"""

from __future__ import annotations

from datetime import UTC, datetime

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.complaint import Complaint

PREFIX = "RW"
SEQUENCE_WIDTH = 6


def next_complaint_number(db: Session, year: int | None = None) -> str:
    year = year or datetime.now(UTC).year
    prefix = f"{PREFIX}-{year}-"

    highest = db.execute(
        select(func.max(Complaint.complaint_number)).where(Complaint.complaint_number.like(f"{prefix}%"))
    ).scalar_one_or_none()

    sequence = 1
    if highest:
        try:
            sequence = int(highest.rsplit("-", 1)[1]) + 1
        except (IndexError, ValueError):
            # A malformed row must not stop new reports; fall back to counting.
            sequence = (
                int(
                    db.execute(
                        select(func.count(Complaint.id)).where(
                            Complaint.complaint_number.like(f"{prefix}%")
                        )
                    ).scalar_one()
                )
                + 1
            )

    return f"{prefix}{sequence:0{SEQUENCE_WIDTH}d}"

File: supabase/functions/api/services/numbering.ts
// Ported from backend/app/services/numbering.py. Human-friendly complaint
// numbers, e.g. "RW-2026-001024". Numbers restart each calendar year.
//
// The Python version derives the next number from the highest one issued
// this year and relies on the table's unique constraint to catch a race
// between two concurrent submissions (retrying on IntegrityError). Postgres
// raises unique-violation as error code 23505 for the same case - callers
// of nextComplaintNumber should catch that from the insert and retry, same
// as the original.

import type { SupabaseClient } from "@supabase/supabase-js";

const PREFIX = "RW";
const SEQUENCE_WIDTH = 6;

export async function nextComplaintNumber(client: SupabaseClient, year?: number): Promise<string> {
  const y = year ?? new Date().getUTCFullYear();
  const prefix = `${PREFIX}-${y}-`;

  const { data, error } = await client
    .from("complaints")
    .select("complaint_number")
    .like("complaint_number", `${prefix}%`)
    .order("complaint_number", { ascending: false })
    .limit(1)
    .maybeSingle();
  if (error) throw error;

  let sequence = 1;
  const highest = data?.complaint_number as string | undefined;
  if (highest) {
    const tail = highest.split("-").at(-1);
    const parsed = tail ? Number.parseInt(tail, 10) : NaN;
    if (!Number.isNaN(parsed)) {
      sequence = parsed + 1;
    } else {
      // A malformed row must not stop new reports; fall back to counting.
      const { count, error: countErr } = await client
        .from("complaints")
        .select("id", { count: "exact", head: true })
        .like("complaint_number", `${prefix}%`);
      if (countErr) throw countErr;
      sequence = (count ?? 0) + 1;
    }
  }

  return `${prefix}${String(sequence).padStart(SEQUENCE_WIDTH, "0")}`;
}
